src/main/java/org/testng/internal/reflect/ReflectionRecipes.java

```diff
@@ -0,0 +1,411 @@
+      boolean firstMethodFiltered = false;
+      final List<Parameter> filterList = new ArrayList<Parameter>(parameters.length);
+      for (final Parameter parameter : parameters) {
+        boolean omit = false;
+        if (currentTestMethod && !firstMethodFiltered) {
+          if (canInject(parameter, InjectableParameter.CURRENT_TEST_METHOD)) {
+            firstMethodFiltered = true;
+            omit = true;
+          }
+        }
+        if (!omit && iTestContext) {
+          omit = canInject(parameter, InjectableParameter.ITEST_CONTEXT);
+        }
+        if (!omit && iTestResult) {
+          omit = canInject(parameter, InjectableParameter.ITEST_RESULT);
+        }
+        if (!omit && xmlTest) {
+          omit = canInject(parameter, InjectableParameter.XML_TEST);
+        }
+
+        if (!omit) {
+          filterList.add(parameter);
+        }
+      }
+      final Parameter[] filteredArray = new Parameter[filterList.size()];
+      return filterList.toArray(filteredArray);
+    } else {
+      return parameters;
+    }
+  }
+
+  /**
+   * Injects appropriate arguments.
+   *
+   * @param parameters      array of parameter instances under question.
+   * @param filters         filters to use.
+   * @param args            user supplied arguments.
+   * @param injectionMethod current test method.
+   * @param context         current test context.
+   * @param testResult      on going test results.
+   * @return injected arguments.
+   */
+  public static Object[] inject(final Parameter[] parameters, final Set<InjectableParameter> filters,
+                                final Object[] args,
+                                final Method injectionMethod,
+                                final ITestContext context,
+                                final ITestResult testResult) {
+    if (filters != null && !filters.isEmpty()) {
+      final ArrayList<Object> arguments = new ArrayList<Object>(args.length);
+      final Queue<ObjectHolder> queue = new ArrayDeque<ObjectHolder>(args.length);
+      for (int i = 0; i < args.length; i++) {
+        // to ensure fifo & hold nulls
+        queue.add(new ObjectHolder(args[i]));
+      }
+      final boolean currentTestMethod = filters.contains(InjectableParameter.CURRENT_TEST_METHOD);
+      final boolean iTestContext = filters.contains(InjectableParameter.ITEST_CONTEXT);
+      final boolean iTestResult = filters.contains(InjectableParameter.ITEST_RESULT);
+      final boolean xmlTest = filters.contains(InjectableParameter.XML_TEST);
+      boolean firstMethodInjected = false;
+      for (final Parameter parameter : parameters) {
+        boolean injected = false;
+        if (currentTestMethod && !firstMethodInjected) {
+          if (canInject(parameter, InjectableParameter.CURRENT_TEST_METHOD)) {
+            arguments.add(injectionMethod);
+            firstMethodInjected = true;
+            injected = true;
+          }
+        }
+
+        if (!injected && iTestContext) {
```

Comment:
Ok! I didn't catch it before. Maybe the loop is easier to read. What do you think @cbeust?